From c4c70fb29c9650573ce7b57243d41e11466cc4d8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Thu, 24 Oct 2024 10:01:34 +0200
Subject: [PATCH] hwrng: stm32 - add RNG support for STM32MP21
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

MP21 needs specific NSCR, HTCR and CR values to work properly.

Change-Id: Ia11ed6cbc87504c8787d66ac0fdf9bb51f5f0eba
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/413763
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/char/hw_random/stm32-rng.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

--- a/drivers/char/hw_random/stm32-rng.c
+++ b/drivers/char/hw_random/stm32-rng.c
@@ -37,6 +37,7 @@
 #define RNG_SR_SECS		BIT(2)
 #define RNG_SR_CEIS		BIT(5)
 #define RNG_SR_SEIS		BIT(6)
+#define RNG_SR_ERROR_MASK	(RNG_SR_CECS | RNG_SR_SECS | RNG_SR_CEIS | RNG_SR_SEIS)
 
 #define RNG_DR			0x08
 
@@ -361,7 +362,7 @@ static int stm32_rng_init(struct hwrng *
 	err = readl_relaxed_poll_timeout_atomic(priv->base + RNG_SR, reg,
 						reg & RNG_SR_DRDY,
 						10, 100000);
-	if (err || (reg & ~RNG_SR_DRDY)) {
+	if (err || (reg & RNG_SR_ERROR_MASK)) {
 		if (priv->bus_clk)
 			clk_disable_unprepare(priv->bus_clk);
 		clk_disable_unprepare(priv->clk);
@@ -520,6 +521,15 @@ static const struct stm32_rng_data stm32
 	.htcr = 0x969D,
 };
 
+static const struct stm32_rng_data stm32mp21_rng_data = {
+	.has_cond_reset = true,
+	.max_clock_rate = 48000000,
+	.nb_clock = 2,
+	.cr = 0x00800D00,
+	.nscr = 0x01F7,
+	.htcr = 0xAAC7,
+};
+
 static const struct stm32_rng_data stm32mp13_rng_data = {
 	.has_cond_reset = true,
 	.max_clock_rate = 48000000,
@@ -541,6 +551,10 @@ static const struct of_device_id stm32_r
 		.data = &stm32mp25_rng_data,
 	},
 	{
+		.compatible = "st,stm32mp21-rng",
+		.data = &stm32mp21_rng_data,
+	},
+	{
 		.compatible = "st,stm32mp13-rng",
 		.data = &stm32mp13_rng_data,
 	},
